text/0000-fn-pin-put.md

```diff
@@ -0,0 +1,189 @@
+This includes the trailing return: a closure containing a `yield` must "return" `!`.
+This will most often be accomplished by a `panic!` or a `loop` at the tail end.
+In most ways (allowed positions, return type), a yield statement acts the same as a return expression.
+
+We call a closure that uses `yield` instead of `return` a **yield closure**.
+
+The key difference is that rather than execution restarting at the top of
+the closure on the next call, execution resumes after the yield statement.
+The compiler will generate a state machine to jump to the correct place upon each resumation.
+
+This is subject to the normal rules about what `Fn` traits are implemented!
+This means that a yield closure can implment all four traits: `Fn`, `FnMut`, `FnPinMut`, and `FnOnce`!
+For example, [`std::iter`'s `Counter` example](counter) can be reimplemented as simply
+
+```rust
+std::iter::from_fn(|| {
+    for count in 1..6 {
+        yield Some(count);
+    }
+    loop {
+        yield None;
+    }
+})
+```
+
+This closure will implement all four `Fn` traits.
+
+So far, however, everything we've seen has been within the domain of the stable `Fn` traits.
+`FnPinMut` is required to support borrowing over `yield` points.
+
+(TODO)
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we *not* do this?
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- Why is this design the best in the space of possible designs?
+- What other designs have been considered and what is the rationale for not choosing them?
+- What is the impact of not doing this?
+
+# Prior art
+[prior-art]: #prior-art
+
+Discuss prior art, both the good and the bad, in relation to this proposal.
+A few examples of what this can include are:
+
+- For language, library, cargo, tools, and compiler proposals: Does this feature exist in other programming languages and what experience have their community had?
+- For community proposals: Is this done by some other community and what were their experiences with it?
+- For other teams: What lessons can we learn from what other communities have done here?
+- Papers: Are there any published papers or great posts that discuss this? If you have some relevant papers to refer to, this can serve as a more detailed theoretical background.
+
+This section is intended to encourage you as an author to think about the lessons from other languages, provide readers of your RFC with a fuller picture.
+If there is no prior art, that is fine - your ideas are interesting to us whether they are brand new or if it is an adaptation from other languages.
+
+Note that while precedent set by other languages is some motivation, it does not on its own motivate an RFC.
+Please also take into consideration that rust sometimes intentionally diverges from common language features.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- Exact naming of `trait FnPinMut`. `FnPinMut` was chosen for this RFC as an
```

Comment:
`FnPin` could also be an option since (as y'all pointed out) `Pin<&mut Self` is the only obviously useful way to take self as pinned.